trie/hexary.py

```diff
@@ -144,6 +162,27 @@ def _traverse(self, root_hash, trie_key) -> Tuple['Node', Tuple[int, ...]]:
 
         return self._traverse_from(root_node, trie_key)
 
+    def traverse_from(self, node_prefix, node, trie_key_input: Tuple[int, ...]):
+        """
+        Find the node at the path of nibbles provided. Unspecified behavior if
+        path ends in the middle of an extension.
+
+        The trie does *not* re-verify the path/hashes from the node prefix to the node.
+        """
+        trie_key = Nibbles(trie_key_input)
+
+        try:
+            node, remaining_key = self._traverse_from(node, trie_key)
+        except MissingTrieNode as exc:
+            raise MissingTrieNode(*exc.args[:3], prefix=node_prefix + exc.prefix)
+
+        # For now, traverse will raise if you try to traverse into
+        # part of an extension node, or leaf node.
+        if len(remaining_key) > 0:
+            raise Exception(f"Tried to traverse {remaining_key!r} past {node!r}")
```

Comment:
TODO: Maybe just return with a remaining key?

I think raising an exception is insufficient. Example: you have explored a trie from one root, knowing that you only need to explore the unmodified parts of the trie. Then the root changes and you want to continue exploring the unexplored parts. But due to possible changes, you might end up in the middle of an extension node on the next traversal.

Reply:
With `traverse_from` it shouldn't be an issue, since you're always navigating from a fixed root (by supplying the parent node).

With `traverse()` though, it's definitely a possibility, and I still need to think through how to handle all the cases properly.

Though... maybe TODO: add root hash to traverse_from() args, to fix the `MissingTrieNode` exception... or make the root hash optional?

Reply:
I'm very likely missing something but did you give any thought to merely returning a representation of the extension node for all paths that don't get you deep enough to hit one of the child nodes?

Reply:
So far, the latest code has an API (direction) that I'm happy with. Traversal is all about knowing where nodes are in the trie, and moving to them. So if you go somewhere that isn't a node (like the middle of an extension or leaf), then it should raise an exception, complete with the information you need to find out what you need to keep traversing (the parent extension node, and the path to that).